### ISSUE TITLE
3D cartesian plot is broken

### ISSUE THREAD

**Issue by ra3xdh** [Qucs/qucs#386]
I have just found another critical bug in Qucs-GUI. 3D-cartesian plot is broken in `master`. Open two schamtics with 3D plots in 0.0.18 and 0.0.19 and compare results to reproduce this bug. 3D plot in 0.0.19 looks unreadable.  Here is example screenshot. Left is Qucs-0.0.18, right is Qucs-0.0.19. 
![3d-diagr](https://cloud.githubusercontent.com/assets/4920080/10156918/583bfc2a-668e-11e5-8177-bed3cb3234e4.png)

**Comment by felix-salfelder** [comment]
it seems, the new code does not print as many cross connections.

the missing blue sample dots in the left picture are more interesting. is this related to the z buffer? what happens if you slightly turn the plot to the left (= move the camera to the right), will they reappear?

please post the data file and the schematic somewhere, i will have a look.

**Comment by ra3xdh** [comment]
Here is test schematic https://gist.github.com/ra3xdh/eb92c81d7d72d401668b/raw/12be21db82ff579e91a9a7e4aa63370a25df2ba3/diode_dblswp_qucs.sch  3D cartesian plot is embedded in schematic. 

> what happens if you slightly turn the plot to the left (= move the camera to the right), will they reappear?

Yes, missing points could appear when plot is rotated at specific angle around Z-axis.

**Comment by felix-salfelder** [comment]
thanks

whether the cross connections are painted or not depends on the angle. must be a bug in the zbuffer code. also, i get a lot of "dangerous: returning negative screen coordinate" messages, which may or may not be related...

**Comment by felix-salfelder** [comment]
apparently, hideLines is broken for 3dplots. i am working on this in #370. previously the z buffer has been monkey-patched into the 2d ploting routines. these have been replaced by the QtPainter.

i think i will implement 3dgraphs with their own drawings...

**Comment by guitorri** [comment]
If at a certain angle the user still can get a useful image, then I will count it as know issue/regression. 
I will move this to the next release, I hope it is ok if you guys.

**Comment by felix-salfelder** [comment]
i have failed to make the old code work again. wasted a lot of time on this `:(`

using qt for the 2d plots is working pretty well. i think we should head for something like qwtplot to do the 3d plots. unfortunately qwtplot does not support multiple layers (yet)...

**Comment by ldpgh** [comment]
Simple check with x=5 SweepPts  :  y=5 SweepPts
* 3D chart placed, hide invisible lines = on, all other settings = default
* The dataset is created (simulated) in 0.0.19 (0.0.9 has an equation issue) and just loaded to the other revisions.
 
Screenshots of the different revisions ... looks like an evolution
* Qucs 0.0.19 ... just a few boundary points left
![qucs_rect3ddiagram_0 0 19](https://cloud.githubusercontent.com/assets/13517069/19830249/fca94aa4-9df3-11e6-8af7-eba1a36130c9.PNG)

* Qucs 0.0.18 ... some visible line segments missing, at least one hidden line segmet visible
![qucs_rect3ddiagram_0 0 18](https://cloud.githubusercontent.com/assets/13517069/19830275/7b36fa9c-9df4-11e6-8ecb-17b2df0b8c0f.PNG)

* Qucs 0.0.16 (2011) ... some visilbe line segments missing
![qucs_rect3ddiagram_0 0 16](https://cloud.githubusercontent.com/assets/13517069/19830276/7edc6d94-9df4-11e6-86e9-61834634ecf3.PNG)

* Qucs 0.0.9 (2006) ... seems to be OK
![qucs_rect3ddiagram_0 0 9](https://cloud.githubusercontent.com/assets/13517069/19830277/862f9a1c-9df4-11e6-92a3-a7c443894c58.PNG)

**Comment by ldpgh** [comment]
Simple check with x= 3 ... 5 SweepPts : y=5 SweepPts
   * 3D chart placed, hide invisible lines = on, all other settings = default
   * It seems the issue has a **strong** correlation to the point count of the x-Axis (1st Sweep)
   * It seems the issue has **no** correlation to the point count of the y-Axis (2nd Sweep)

Screenshots using 0.0.19 and different point size of the x-Sweep
* 2pts ... makes less sense ... it's simply flat

* 3 pts
![qucs_rect3ddiagram_0 0 19__3pts](https://cloud.githubusercontent.com/assets/13517069/19830306/4baffc96-9df5-11e6-8e18-86c47a83da65.PNG)

* 4 pts
![qucs_rect3ddiagram_0 0 19__4pts](https://cloud.githubusercontent.com/assets/13517069/19830309/508641e4-9df5-11e6-9a9b-c48e2b201e15.PNG)

* 5pts ... same as the screenshot in the version related comment above
![qucs_rect3ddiagram_0 0 19__5pts](https://cloud.githubusercontent.com/assets/13517069/19830310/54670b5e-9df5-11e6-9185-beef5048207c.PNG)

**Comment by ldpgh** [comment]
Is there any Unit-test associated to rect3ddiagram.cpp?

**Comment by felix-salfelder** [comment]
i remember we had been discussing unit tests for the GUI, something about graphical diffs and they are somewhere on a whishlist.
@ldpgh do you know how to implement that? please don't hesitate!

**Comment by ldpgh** [comment]
Environment:  Win7, qucs-HEAD, CMake in QtCreator 3.1.2 @ Qt4.8.6, ming482.

There is a significant difference of appearance between debug-mode (significant more line segments visible in 0.0.19, regardsless breakpoints)  vs. non-debug-mode.

Have you experienced similar behavior?

**Comment by felix-salfelder** [comment]
as you might have guessed, the debug mode must not interfere with the behaviour of the programme (except for printout to stderr), so that's another bug.

**Comment by in3otd** [comment]
...interesting, maybe some floating point comparisons are giving different results, since the variables may end in registers instead of memory. I think we saw something similar in the simulator in the past.

Doing a quick diff of `rect3ddiagram.cpp` between version `0.0.6` and the current one doesn't appear to show any big changes (besides some slightly different rounding/truncation), at a first glance, but I may well have overlooked something.

**Comment by ldpgh** [comment]
@in3otd ... current version vs version 0.0.6 ... Do you mean Qucs version 0.0.6, because for `rect3ddiagram.cpp `it looks to me there are 2 major implementation changes:
- some data types involved are change from `int `to float
- from 1-dim pointer array (p, p+1) to pointer array with points (p.x, p.y) including change of method definitions
In addition it seems some changes are caused by graph.cpp/.h respective lead to changes in graph.h/.cpp.

**Comment by ldpgh** [comment]
Not verified/debugged yet . . . just guessing (!!!) . . . will check it next day.
The differences smells like "malloc without memset" (e.g. `removeHiddenLines()`)Some strange behavior seems to be related how the application is started too..

What is the more appropriate direction taking into accound all the side effects
1. fix current revision with all the floats
2. lift up the implementation of 0.0.9 (which sounds gracy) to match current 0.0.19 implementation of graph.h/.cpp and diagram.h/.cpp

findings-1
with `hidden `enabled and return value of `isHidden()` overriden to `true` all drawn line segments look OK ... means same as `hidden=false`. This is surprising to me, because I have not just missing segments. Sometimes line segmets start/end outside the diagram-area, which would point to the framebuffer and the associated files.

**Comment by ldpgh** [comment]
(1) Start in Run-Mode (Ctrl-R) ...

(2) Load rect3ddiagram1x_test_sch.sch
![qucs_rect3ddiagram_0 0 19__one_kind](https://cloud.githubusercontent.com/assets/13517069/19941160/53dc6386-a12f-11e6-80dd-a4102aaa135a.PNG)

(3.1) Load rect3ddiagram7x_test_sch.sch
(3.2) Close Load rect3ddiagram7x_test_sch.sch
(3.3) Close Load rect3ddiagram1x_test_sch.sch

(4) Load rect3ddiagram1x_test_sch.sch
![qucs_rect3ddiagram_0 0 19__other_kind](https://cloud.githubusercontent.com/assets/13517069/19941164/59bd6002-a12f-11e6-9e8f-6c5aa87e0f53.PNG)

Note-1: Closing order @ 3.2/3.3 has impact : 7x before 1x
Note-2: Hide/Unhide of SelectionWindow has impact too !

**Comment by ldpgh** [comment]
findings-2
with `hidden `enabled and return value of `isHidden()` overriden to `true` all drawn line segments look OK ... means looks the same as `hidden=false`.
This is surprising to me, because I have not just missing segments. Sometimes line segmets start/end outside the diagram-area, which would point to the framebuffer and associated files.
![qucs_rect3ddiagram_0 0 19__outlier](https://cloud.githubusercontent.com/assets/13517069/19939671/3f24749c-a12a-11e6-8251-489bb6b2d94c.PNG)

**Comment by felix-salfelder** [comment]
... also try to plot 3D data (i.e. sweep over 3 parameters). that gives a rough idea of what the code is trying to do. then 4.

**Comment by ldpgh** [comment]
It's cool stuff (with hideLine=true) ... for this reason it make sense to spend some time with it.
![qucs_rect3ddiagram_0 0 9__3waves](https://cloud.githubusercontent.com/assets/13517069/19947087/d6eed35a-a146-11e6-8205-7aef70edfc17.PNG)

And a chart with all line segments visible for comparison.
![qucs_rect3ddiagram_0 0 9__3waves_nohide](https://cloud.githubusercontent.com/assets/13517069/19947097/defedff4-a146-11e6-9d1b-14d7f3a6d764.PNG)

Both pictures added just to see the goal :-)

And a screenshot of the current status (0.0.19 @ Oct 2016) ...
![qucs_rect3ddiagram_0 0 19__3waves](https://cloud.githubusercontent.com/assets/13517069/19947237/7e5d5e18-a147-11e6-8f1e-5f11edd446dd.PNG)

**Comment by ldpgh** [comment]
I really like to get a feedback regarding appropriate direction . . . taking into accound all the side effects

1. fix current revision with all the floats
2. lift up the implementation of 0.0.9 (which sounds gracy) to match current 0.0.19 implementation of graph.h/.cpp and diagram.h/.cpp

It guides my debug strategy/procedure. Keep going with both would require at least one more head ;-)

**Comment by felix-salfelder** [comment]
the reason why i rewrote parts of the plotting (mostly unmerged work) is, that i would like to read data from other sources than just .dat files. think of hdf5 or a simulator as a possible alternative souce (for now). the dat format is absolutely unsuited for this type of task, but unfortunately, the plotting routines more or less work on a binary image that is layed out like these .dat files... which is where the shit is hitting the fan.

that said, when i tried to fix the 3d plot (yes, i'd better be able to, but i wasnt), it was pretty clear to me that this would the wrong track. it never worked right (and not only the comments indicate that it is trial-and-error-spaghetticode). one part of this desaster is, that the 3dplot and the 2dplot are heavily entangled, and they should not be. if you can untangle it, then yes, lift 0.0.9. but *please* only for the 3d plotting. if you prefer a sane path, please help with #370, it fixes some other problems with the markers as well...

please let me know, if you need more info. will have more time for this after the symbols are in place...

**Comment by ldpgh** [comment]
... continue debug with HEAD (9d924caa3879fbca978987a9fe964eadbc5f0fc4)

rect3ddiagram.cpp : isHidden()  + hideLines=true (!)
- return 1==0;
- all line segments are drawn. It seems graph.cpp/h & diagram.cpp/h are not involved respective do not modify what has to be drawn

rect3ddiagram.cpp : removeHiddenLines() + hideLines=true (!)
- free(zMem); return; @ line 496 ... all line segments are drawn
- free(zMem); return; @ line 549 ... reduced output as already known (just few line segments are drawn)
- it seems to confirm assumption regarding graph.cpp/h & diagram.cpp/h

**Comment by ldpgh** [comment]
Finding-4
rect3ddiagram.cpp : calcCoordinate3D(...) + hideLines=true (!)
```
p->x  = int(calcX_2D(x, y, zr) + 0.5 + xorig);
p->y  = int(calcY_2D(x, y, zr) + 0.5 + yorig);
```
 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ^^^
Remove of `0.5` brings a couple of line segments back, but it seems to be just a sideeffect. Reopen of the schematic brings sometimes the old status back.

**Comment by ldpgh** [comment]
As Felix wrote ...the algorithm within the source code is difficult to track. 

2 modifications in rect3ddiagram.cpp (5 lines, other would say hacks and I agree) improve the behavior ... just improved ... not fixed. And the random behavior seem to be still their ... maybe caused by my setup.

Screenshot of the 0.0.19Mod @ 8th Nov 2016
![qucs_rect3ddiagram_0 0 19mod__3waves](https://cloud.githubusercontent.com/assets/13517069/20115173/9240a656-a5f7-11e6-9e44-42c2a43c9f20.PNG)

Screenshot of the 0.0.19 @ Oct 2016
![qucs_rect3ddiagram_0 0 19__3waves](https://cloud.githubusercontent.com/assets/13517069/20115204/aaa25b86-a5f7-11e6-9e66-0df25b7b538d.PNG)

Would like to have someone else to check it.

There are a schematics (see above) within my test project ... maybe qucs-test is a good place for it as long as rect3ddiagram is part of the game.

Who can help?

**Comment by ldpgh** [comment]
Finding-5
rect3ddiagram.cpp : removeHiddenLines(...) . . . segmentation fault

In order to chase the random behavior the initial` malloc()` for `zMem `& `Mem `has been changed from 
`malloc( (Size+2)*sizeof(...))` -> `malloc( 8*(Size+2)*sizeof(...))` ... just to prevent the `realloc()`. No more lines leaving the diagram area.

Adding test-case diode_dblswp_qucs.sch with 200pts @ SW1 has brought up a memory issue (segmentation fault at `fprintf ` for debugging respective @ `calcLine : pMem->x = ...`) while turning the malloc back to the original settings (remove of `8*`).

With the modified setting (`8*(...)`) diode_dblswp_qucs.sch runs with 20000pts @ SW1 and 50pts @ SW2 with no issues.

**Comment by ldpgh** [comment]
0.0.19Mod @ 13th Nov 2016 ... modification are in rect3ddiagram.cpp

Comparison ... 0.0.9 ... 0.0.19 ... 0.0.19 Mod ... and how it looks at my screen.
![cmp_revs_small_000](https://cloud.githubusercontent.com/assets/13517069/20249610/3a6963d2-a9fd-11e6-91b2-5da686030e4b.png)
![cmp_revs_small_test](https://cloud.githubusercontent.com/assets/13517069/20249612/3ca7e254-a9fd-11e6-8e75-e52a8c0e6013.png)
![cmp_revs_test_sch](https://cloud.githubusercontent.com/assets/13517069/20249614/3f8b0e10-a9fd-11e6-92b7-67117e0e8d2f.png)
![cmp_revs_dio](https://cloud.githubusercontent.com/assets/13517069/20249608/37342224-a9fd-11e6-91ad-add4e40bbed8.png)
![cmp_revs_waves3x_sweeps4x](https://cloud.githubusercontent.com/assets/13517069/20249615/4205b21c-a9fd-11e6-9844-c715855f509f.png)
![cmp_revs_waves3x_test](https://cloud.githubusercontent.com/assets/13517069/20249616/4536d0b0-a9fd-11e6-925f-de3ba3775124.png)

**Comment by felix-salfelder** [comment]
wow. care to open a PR with the modified recd3dthing.cpp?

**Comment by guitorri** [comment]
@ldpgh can you share your fix? If you do we can include it already on 0.0.19 which is about to be released.

**Comment by ldpgh** [comment]
@ guitorri & Felix ... "share the files" is the plan.

Understand Git/Github PR-procedure is still a hurdle. And I'm still fiddling with my debug comments spread over the entire file ... to a large degree no longer valid.

Is there any deadline? My plan is to get it done until end of this week.

**Comment by guitorri** [comment]
I tentatively put December 1st for the release. So take your time and let us know if you need any assistance. This might help: https://github.com/Qucs/qucs/wiki/Contribution. In particular the workflow section.

**Comment by felix-salfelder** [comment]
> debug comments spread over the entire file

`git add -p` might be useful. rescued me a couple of times (out of similar situations).

**Comment by guitorri** [comment]
Fix merged into `release-0.0.19`. Thank you ldpgh.
